### DISCUSSION TITLE
2024-11-07 - Developer Secrets - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by GordonBeeming** [SSWConsulting/SSW.Rules.Content#7137] [Chewing the Fat]
Hey Team,

Let's talk about storing and sharing developer secrets
https://www.ssw.com.au/rules/share-your-developer-secrets-securely/

1. Do you have any opinions on how your project should store and share Developer secrets?
2. Any horror stories?
Tip: Probably best not to mention customer names.

ℹ️ Note: We are now using Giscus, so reactions and comments on the GitHub Discussion will appear on the rule

**Comment by leahy268** [comment]
>Do you have any opinions on how your project should store and share Developer secrets?

Yes, Keeper, Keeper Keeper...  I have another video on this topic.. also not Lastpass with its current reputation.

One other thing is that they have a service that links directly to GitHub and Azure Devops that can get secrets directly from Keeper as well as rotate service account passwords and all kinds of things.  Currently we aren't using it though :(


>Any horror stories?

Many but the most recent was a current client. One of their Devs offered to send me the secrets via Teams..  I had to say, don't send them in plain text via chat. The response was well how else can I.

**Comment by jakebayliss** [comment]
Love using Keeper for this :)

**Comment by TobyChurches** [comment]
Keeper is Amazing for storing secrets that are going to be used regularly! But if its just a one off, I really like using 1Ty.me

**Reply by leahy268** [reply]
Agreed.  Except that the other disadvantage of 1Ty.me and other services like that is that we don't know for certain how secure they are.

**Comment by calumjs** [comment]
1ty.me is my preference for a quick and easy way to share secrets.
I do always cringe when someone puts the entire connection string, or domain, username, password, etc. in there because you are basically trusting that they don't peek at the data, don't have a rogue dev on staff, etc.
Do we even know the website isn't run by North Korean hackers?
However, if you just put the password in there with no additional context, I'd say it's pretty safe. You can't do much with just a password and no idea where to use it.

**Comment by StefStar** [comment]
I am not a devloper but definitely understand the need to be careful and precise with passwords in marketing! No horror stories yet of passwords getting leaked but always scared of the possibilities....

**Comment by joshbermanssw** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

Keeper is great for this! Used it with all my projects so far... sometimes storing some secrets/ env variables in vercel too

> Any horror stories?

Not yet ... 👀

**Comment by ricksu978** [comment]
Always Keeper or Azure Key Vault!

**Comment by Ben0189** [comment]
Using Azure Key vault and Keeper all the time 😊, in some use case i use managed identity (don't need any secret)

**Comment by steven0x51** [comment]
Keepere and Azure Key Vault are awesome

**Comment by isaaclombardssw** [comment]
Glad we're recommending keeper – I like the tool 👍

**Comment by kikibianc** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?
Keeper is the way to go.

> Any horror stories?
No horror stories 👀

**Comment by Calinator444** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

endpoints should ideally be secured using Azure key vault or bearer tokens. For CI/CD pipelines it's best to secure secrets using GitHub Secrets (If using GitHub).

> Any horror stories?

No

**Comment by awaltos** [comment]
Keeper and Key Vault or similar cloud secret storage services like AWS KMS, etc.
No horror stories yet 🤞️ 🤜️🪵️

**Comment by jimmy1992abc** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

The Keeper currently running very well.

> Any horror stories?

Not yet

**Comment by bahjatalaadelssw** [comment]
Keeper is awesome. you. can organise access and add detailed instructions.

**Comment by drwharris** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

All config should be in App Configuration including secrets as a reference to KeyVault. This way, the only thing you need is permissions on App Config. This way 

> Any horror stories?

Trying to store config in KeyVault is not a great idea!

**Comment by RobJT-wq** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

Keeper or another Enterprise Secret Management Tool is the best option for sharing secrets

> Any horror stories?

None so far

**Comment by Lewkans** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

Keeper works well, and allows you to store many other types of secrets as well, which is also a nice bonus 

> Any horror stories?

None at the moment

**Comment by jeoffreyfischer** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

K-E-E-P-E-R

> Any horror stories?

Not yet 🫠

**Comment by ncn-ssw** [comment]
> 1. Do you have any opinions on how your project should store and share Developer secrets?

I'm wary of 1Ty.me - as @leahy268 has mentioned, it is hard to tell how secure it actually is.

> 2. Any horror stories?

Not yet.

**Comment by tkapa** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

Keeper has been very easy to use and access, the only danger is that they can become out of date, like any documentation

Totally the best method here - https://www.youtube.com/shorts/Udf44K6rt-E

> Any horror stories?

I think at my first job, secrets were stored in the repo

**Comment by chrisschultzssw** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

As @leahy268 mentioned, Keeper has a dedicated Secrets Manager that integrates with GitHub, AzDO, etc.
https://www.keepersecurity.com/secrets-manager.html

> Any horror stories?

I've seen too many people (not at SSW) saving passwords in notepad/excel/on paper...

**Comment by vladislav-kir** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

Keeper is amazing :)

> Any horror stories?

No

**Comment by Tanya-Leahy** [comment]
I'm not a developer, but I fully appreciate the need to handle passwords with care in the workplace. No horror stories of leaked passwords yet.

**Comment by sam-s-maher** [comment]
Keeper as it is also a sort of documentation in itself, when the secret updates for one it updates for all. Additionally, in a perfect world where all secrets are there and well named, there's no need for the overhead of asking devs to fetch them for you.

**Comment by Marxoz** [comment]
Azure Key vault and Keeper does the job for me!

**Comment by KostaMadorsky** [comment]
I use 1Password, which allows you to share an item with expiration, one-time open, and restriction to specific emails (they do not need to have 1Password). You can also see the history of your shares (if they were opened, etc.) and cancel non-expired ones.

![image](https://github.com/user-attachments/assets/5fcb81eb-7f38-43e8-b83f-73da9e86e4b5)

**Comment by bettybondoc** [comment]
I’m a fan of tools like Keeper and 1Password for managing passwords and secrets.

👻 Horror Story: Years ago, I remember an incident where one of our devs accidentally pushed an entire .env file with all our API keys, passwords, and DB connection strings to a public GitHub repo. I _think_ they forgot to setup .gitignore. I had just joined the team then so I wasn't part of the cleanup crew but our seniors were busy for a few days rotating all keys, scrubbing the repo, and writing incident reports/docs.

**Comment by amankumarrr** [comment]
If I'm sharing secrets with an application, Key Vault would definitely be my first choice. Otherwise, I'd go with Keeper or 1tyme.

**Comment by stephanfakossw** [comment]
N/a not a dev but sharing logins via Keeper

**Comment by AttackOnMorty** [comment]
Have we considered 1Password Enterprise? 😀

<img width="1160" alt="image" src="https://github.com/user-attachments/assets/3f5da815-815a-4f70-9f29-f8d45bca2a64">

**Figure: 1Password is more popular**

**Comment by PennyWalker** [comment]
Great video Jeoffrey! Really enjoyable to watch.

I love Keeper! It's so easy to use and I love that I can have it on my phone :)

**Comment by miichaelsmedley** [comment]
> 1. Do you have any opinions on how your project should store and share Developer secrets?

Azure Key Vault

> 2. Any horror stories?

Not dev secrets but 2nd hand issues with plain text in a URL string...

**Comment by SamBWagner** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

I like KeyVault and I time links. That'll do for me :)

> Any horror stories?

I've done the goofy thing of committing secrets to a repo. Ya do it once in your life. Just make it not hurt haha

**Comment by jimmidier** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

For most projects, cloud services like Azure Key Vault are a good option. In an air-gapped environment, a self-hosted secret store is ideal, but if that's not available, a shared drive can also work.

> Any horror stories?

Although I've seen secrets checked into source I haven't come across anything serious.

**Comment by thom-wang** [comment]
Loved keeper, really handy tool

**Comment by babakamyljanovssw** [comment]
I've been using Keeper, Azure KeyVault, and sometimes 1ty.me. I think for sharing dev secrets Keeper was best so far.

**Comment by jerryluossw** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

I think keeper or similar tools are cool.

 > Any horror stories?

I accidentally pushed my own Alibaba Cloud database connection string to my open-source project many years ago. In the end, I had to delete the repository. Since then, I've been especially careful in managing keys for open-source projects.

**Comment by danielmackay** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

IMHO using 1ty.me is NOT a good way to share secrets. We are entrusting our most valuable data to a random company that we are not paying. When you're not paying with money, you're paying with your data.

Instead we should be using the one-time sharing that's built into Keeper:

<img width="511" alt="image" src="https://github.com/user-attachments/assets/dac786bb-d371-44c2-8c7f-0e4341b3ef96">


> Any horror stories?

I've seen clients that checked in files named `dev.secrets.json` into source control. They didn't even try to hide it! 😱

**Comment by JackDevAU** [comment]
> 1.Do you have any opinions on how your project should store and share Developer secrets?

Keeper is great and everyone has access to it 🚀

>2. Any horror stories?

Client was storing all username (email) + password as plaintext in their database 😱 What made it worse, was the data that you could get would not be something that you would want leaked... 😅

**Comment by ozair-ashfaque** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

Keeper

> Any horror stories?
Tip: Probably best not to mention customer names.

Secrets were committed to the repo.

**Comment by matt-goldman-ssw** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

User secrets + Keeper - I thought we all knew this already?

> Any horror stories?

I've seen secrets in the repo wiki, directly in code, etc.

**Comment by tino-liu** [comment]
> Do you have any opinions on how your project should store and share Developer secrets?

Keeper. It's so easy and nice to use.

> Any horror stories?

Hard coding user secrets! That's a nightmare ha.

**Comment by Aibono1225** [comment]
Love Keeper and Keeper❤
